src/transformer_util.ts

```diff
@@ -0,0 +1,545 @@
+
+    const nodePath: ts.Node[] = [];
+    visitNode(sourceFile);
+    return sourceFile;
+
+    function visitNode(node: ts.Node) {
+      const startNode = node;
+      const parent = nodePath[nodePath.length - 1];
+
+      if (node.flags & ts.NodeFlags.Synthesized) {
+        // Set `parent` for synthetic nodes as well,
+        // as otherwise the TS emit will crash for decorators.
+        // Note: don't update the `parent` of original nodes, as:
+        // 1) we don't want to change them at all
+        // 2) TS emit becomes errorneous in some cases if we add a synthetic parent.
+        node.parent = parent;
+      }
+      fileCtx.syntheticNodeParents.set(node, parent);
+
+      const originalNode = ts.getOriginalNode(node);
+      // Needed so that e.g. `module { ... }` prints the variable statement
+      // before the closure.
+      // tslint:disable-next-line:no-any
+      (node as any).symbol = (originalNode as any).symbol;
+
+      if (originalNode && node.kind === ts.SyntaxKind.ExportDeclaration) {
+        const originalEd = originalNode as ts.ExportDeclaration;
+        const ed = node as ts.ExportDeclaration;
+        if (!!originalEd.exportClause !== !!ed.exportClause) {
+          // Tsickle changes `export * ...` into named exports.
+          // In this case, don't set the original node for the ExportDeclaration
+          // as otherwise TypeScript gets confused.
+          ts.setOriginalNode(node, undefined);
+        }
+      }
+
+      let importOrReexportNode: ts.Node&{moduleSpecifier: ts.StringLiteral}|undefined;
+      if (node.kind === ts.SyntaxKind.ImportDeclaration) {
+        const id = node as ts.ImportDeclaration;
+        // this should always be a StringLiteral, see the typescript docs.
+        if (id.moduleSpecifier.kind !== ts.SyntaxKind.StringLiteral) {
+          throw new Error(`Unexpected moduleSpecifier kind: ${id.moduleSpecifier.kind}`);
+        }
+        importOrReexportNode = node as ts.Node & {moduleSpecifier: ts.StringLiteral};
+      } else if (node.kind === ts.SyntaxKind.ExportDeclaration) {
+        const ed = node as ts.ExportDeclaration;
+        if (ed.moduleSpecifier) {
+          // this should always be a StringLiteral, see the typescript docs.
+          if (ed.moduleSpecifier.kind !== ts.SyntaxKind.StringLiteral) {
+            throw new Error(`Unexpected moduleSpecifier kind: ${ed.moduleSpecifier.kind}`);
+          }
+          importOrReexportNode = ed as ts.Node as ts.Node & {moduleSpecifier: ts.StringLiteral};
+        }
+      }
+      if (importOrReexportNode) {
+        fileCtx.importOrReexportDeclarations.push(importOrReexportNode);
+      }
+
+      // recurse
+      nodePath.push(node);
+      node.forEachChild(visitNode);
+      nodePath.pop();
+    }
+  };
+}
+
+/**
+ * Transform that needs to be executed after TypeScript's transform.
+ *
+ * This fixes places where the TypeScript transformer does not
+ * emit synthetic comments.
```

Comment:
Yes. I think I will leave TODOs everywhere for now where I need to file a bug, as that takes a bit (with small reproduction, ...)